UI safety updates rebranch

flammableliquids opened the pull request:
# Description

The forms "Ripper Settings" And "UI Settings" Have now been upgraded.
This includes the beginnings of field validatiors and drop downs.
"Ripper Settings" form has too many fields to be defined manually in python, and so a json has been created to allow for ease of organization and management of the form. 

Fixes # (issue title here)

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Many many repeated tests in 'bare metal mode'

- [ ] Docker
- [x] Other (24.04 Ubuntu)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have tested that my fix is effective or that my feature works

# Changelog:

Most recent fixes and improvements:
- write to arm.yaml logging has been improved.
- - better logging
- - cleaner data-typing for json to yaml
- commented debug messages have been removed.
- custom validators for dynamically created forms work best when the validators are classes not functions.
- dynamic validator creation has been moved into functions outside of class creation, making code easier to read.
- testing successful.

**I am still trying to figure out how to modify the jinja to show the validation errors.**
**It will now prevent people from saving dangerous or wrong configurations, by erroring to a log file,**
**_and giving no visual feedback to user._**
(I'm still trying to figure out how to format the form in jinja macros to have the error come up, and the users focus re-directed.
I felt it best to make the pr to prevent people from screwing up)


- arm/ui/settings.SettingsForm() is not hard-coded for rendering, if a new field is added to the form, the loop will now render it.
- applied to "UI Settings":
- - DataRequired validator now 
- - Hinting now applied to fields (not just the blue 'i' icon but on hover.
- - use_icons and save_remote_images changed to dropdowns to prevent user mistakes.
- arm/ui/forms.SettngsForm() completely re-created:
- - is no longer a class, but a function that returns an instance of a class.
- - creates an instance an empty class, and then based on ripperFormConfig.json populates the fields(types)
- - Hinting now applied to fields.
- - Dropdown fields and integer fields now enforce user input to prevent users from crashing their servers (This all started because I typed DEBRUG instead of DEBUG and had to plug in to the server to find the mistake). 
- ripperFormConfig.json created, and quickJsonGenerator.py (to create ripperFormConfig.json), so that the field types, and validators for the "Ripper Settings" form can be easily managed via loops instead of massive python code.
- arm/ui/forms_custom_validators.py added. To hold custom validators. Primary goal validator is the "validate_path_exists" to prevent users from inputting files or folders that do not exist on the server. (If you do this, you will cause your server to crash) (validator either does not work, or the rendering of form error fields doesn't show the error)
- improved datatyping in forms.py
- arm/ui/utils.py Updated: Improved dataype indications and function return type indications. Functions modified or added:
- - listCoPairedIntoTuple (Ability in name).
- - jsonFile_to_dict : provided a str|Path will read the file and return a dict.
- - generate_comments modified to use jsonFile_to_dict
- - generate_ripperFormSettings reads ripperSEttingsConfigFile.json and uses jsonFile_to_dict to return a dict
- Jinja Changes:
- - _form_field.html created, containing the marco 'render_field(field)', takes a field, and renders it, including the 'i' popover.
- - arm/ui/settings/templates/settings/ui.html imports render_field,  and instead of manually executing against each field in the form, loops over the passed form, and calls render_feild
- - arm/ui/settings/templates/settings/ripper.html imports render_field,  and instead of manually executing against each field in the form, loops over the passed form, and calls render_feild
- [BugFix] arm/runui.py updated for better path support.
- [BugFix] arm/ui/comments.json was Missing the UMASK Field
- README Fix Including link correction
- dynamic form creation has been made more modular. Making it easier to read, and reusable for when other forms are made dynamic.
- extraneous comments have been removed

SonarCubeCloud will generate a security warning, because I am using Jinja to render fields in HTML, considering that this project is not meant for 'production web' implementations, I am willing to accept the unlikely security risk

flammableliquids committed:
Update runui.py

sys append path was resolving to null.
Fixed to resolve to the correct parent directory

flammableliquids committed:
Update README.md

Typo in ReadMe fixed. Links to correct page.

flammableliquids committed:
Update DebianInstaller.sh

Correct folder

flammableliquids committed:
Prototyping UI Form input safety

As a foundation to prevent users from inputting bad configuration and crashing the server, for input validation needs to be enabled.

An initial attempt was made to use a validation function, but this validation is not utilized as the current implementation of form rendering is not quite WTF Form standard.

Settings Page Ui Settings Form has now been successfully prototyped as something more WTF Form standard. SelectFields at definition of Form are now respected.

ui.html has been cut down significantly, and mostly replaced by calls to _for_field.html based render macros.

ui/settings/settings.py:settings renders the whole page at once.  Defining the information populated in the FlaskForm Object here, allows it to be handed off to the render, and still have form submit working.
Help icon renders. but getting tooltips and json based comments is still outstanding

flammableliquids committed:
- generation of comment dictionary is moved earlier.
- a loop over a FlaskForm's fields is used to populate teh data and .data for the field. This allows for dynamic changes to db fields to populate the 'existing' values of a form. If a new field is required.
- this method should allow for other forms to be dynamically populated with exsting db values.
- jsoncomments is not passed to the jinja render for ui_settings, as the render_kw title is now set during the loop.
- tooltip's or fields is now populated, along with the help field
- raw field_name is visible in the 'help' icon. As a more user friendly name can be populated in a form label

flammableliquids committed:
ui.html will now loop over all fields in a form that do not match submit or csrf_token, and submit them to the render macro.
no longer needing to specify which field to render in which order.

flammableliquids committed:
nicer titles added to UiSettingsForm

flammableliquids committed:
README.md updated with corrected links

flammableliquids committed:
working prototype of ripper settings being built directly from arm.yaml configuration file.

flammableliquids committed:
ripper settings form, fixes

Help comments now populate, not only with the 'i' icon, but also as tooltips on the fields.

Saving the form now works properly, as the 'POST' functino save_settings, now creates an instance of the form object, but populated based on request.form.to_dict().

flammableliquids committed:
- new form validation function, which validates if umasks for chmod command to see if they are valid.
- SettingsForm() function modified to read ripperSettingsConfig, and then use that to establish what field types are required, and their validators.
- - needs modification to funtion with comments.json
-  quickJsonGenerator generates ripperFormCOnfig based on arm.yaml and comments.json, not intended for production use, just to regenerate the ripperFormCOnfig
- ripperFormConfig.json is intended to be a list of all fields in the ripper form, and the related field types and their validations.
- - Comments.json will still be used to define the comments (keeping that central)
- - default value, and comment shoudl probably be removed from the field for continuity. They are included in the file at generation to make it easier to define validations and field types.
* this commit isn't working 100%, it is being made to ensure that I have a rollbackpoint during develop *

flammableliquids committed:
Added UMASK Hint

flammableliquids committed:
- an attempt at custom field validation to protect against stupid user input
- fixes to SettingsForm class.
-- cleaner dict references.
-- when a form is created it pull's it's own of comments.
-- moving away from dropdowns to radials (Maybe easier on a phone to have a button instead of a dropdown.
-- thinking about IntegerRange fields for sliders.

flammableliquids committed:
loop to populate the values in the form, before rendering.

flammableliquids committed:
not intended for public usage. Just a quick and dirty script to mash together a ripperFormConfig

flammableliquids committed:
Name and field names are designed to clearly communicate.

flammableliquids committed:
- Function created to return a dict for configuring a ripperForm
- repeatable json file to dict function.
Which is then called to read ripperFormConfig and comments.json

flammableliquids committed:
First attempt to get Radio Buttons to render in  a nicer way.

flammableliquids committed:
- Clearer typing
- funciton to return a list of tuples added

flammableliquids committed:
SelectField Renders better than RadioField for the moment. Fix the save, then work on the asthetics.

flammableliquids committed:
- undeeded import removed (more to come)
- list comprehension moved to function for nicer typing validation.
- improved funciton commenting.

flammableliquids committed:
moved debug logging message inside teh validate_on_submit function.

and I even tried to wrap the write to file in an try catch, to give some indication of why it's not working

flammableliquids committed:
Custom validators appear to not be working properly. Either not throwing the error, or the error isn't rendering. So I removed them until I can figure out why
-  saving the settings still doesn't work properly. The file changes timestamp, but the values don't appear to write.

flammableliquids committed:
Added function datatype recommendations to make intellisense work better, and help fault find for easier clarity.
ripperFormConfig.json mistakes removed.

flammableliquids committed:
custom validators have been moved into a separate file as it feels cleaner this way.

flammableliquids committed:
form custom vaildators move to different file.
RipperSettings no longer tries to set default values, that caused complexity problems when trying to open an existign config file and save it.

flammableliquids committed:
getting arm.yaml configuration into the form object seems to work better if you loop over the cfg.arm_config.items, rather than looping over form._fields.items, maybe because it's a dynamic form? Unsure, but likely as submit is the only default field. But it works.

flammableliquids committed:
Merge b3cca74f770520c51d9a1dc253fdbf15abbc57cb into afad2ef3fe7bd1edce42607ea6b91e4613ab1222

github-actions[bot] committed:
[Automated] Release: minor - Version from 2.17.4 to 2.18.0

flammableliquids committed:
Merge f00bf1ba8d1790375be3ab0346a062401c46e22f into afad2ef3fe7bd1edce42607ea6b91e4613ab1222

github-actions[bot] committed:
[Automated] Release: minor - Version from 2.18.0 to 2.19.0

flammableliquids committed:
the 'to do file' of _form_field_radials rendering seems to annoy the AI code validation, so I'm removing it.

flammableliquids committed:
Merge a3cddbbd78b58183ce75d498d64bd449f554fe5a into afad2ef3fe7bd1edce42607ea6b91e4613ab1222

github-actions[bot] committed:
[Automated] Release: minor - Version from 2.19.0 to 2.20.0

flammableliquids committed:
umask logic validation fix.
the error message isn't completely accurate, but the logic is now fixed.

flammableliquids committed:
Merge 3bd66266f9ccd127df3e6920242f81e4455da717 into afad2ef3fe7bd1edce42607ea6b91e4613ab1222

github-actions[bot] committed:
[Automated] Release: minor - Version from 2.20.0 to 2.21.0

flammableliquids committed:
Merge branch 'main' into UI-Safety-Updates-Rebranch

flammableliquids committed:
Merge 7e7f5409b0bdc739f4d6a07c8668ce3b55416c33 into edf25a496bf388228df001be7f1f0d76524b1a32

github-actions[bot] committed:
[Automated] Release: minor - Version from 2.21.0 to 2.22.0

flammableliquids committed:
Update VERSION

version fix

microtechno9000 committed:
Merge branch 'main' into UI-Safety-Updates-Rebranch

flammableliquids committed:
Merge branch 'UI-Safety-Updates-Rebranch' of https://github.com/automatic-ripping-machine/automatic-ripping-machine into UI-Safety-Updates-Rebranch

flammableliquids committed:
Import error fix, and recommendations taken

- forms.py added missing import to align with default branch.

- arm.ui.utils.jsonFile_to_dict adjusted upon recommendation, no longer nested, solved by 'not' and early return.

-  arm.ui.settings.settings.save_settings .close() removed.

flammableliquids committed:
Merge remote-tracking branch 'origin/main' into UI-Safety-Updates-Rebranch

flammableliquids committed:
Stype and formatting changes flake8 and ARM UI Tests

flammableliquids committed:
invalid escape sequence and spacing

flammableliquids committed:
visual indents and other linting completed

flammableliquids committed:
typing and formatting

flammableliquids committed:
linting SettingsForm attempt at simplificiton

- Extraneous spaces have been removed.

- Extra carriage returns have been added.

- creation of the field object for a form has now been moved into another two functions, to reduce the apparent complexity of the function, also allowing for reuse in other forms.

flammableliquids committed:
more linting

de-linting

flammableliquids committed:
Update forms.py

flammableliquids committed:
- attempting to figure out why "save_settings" no longer works.
 - added new datatype checkers.
 - improved and cleaned up debug level logging
-

flammableliquids committed:
- forms: removed duplicate lines.
- overhauled arm.yaml config generation to use better datatype checks, string wrapping and character checking. (should run faster too), considering importing the form configuration to use as a datatype conversion reference.

flammableliquids committed:
- added failure check for save_ui_settings
- moved reload of logging level outside of try catch

flammableliquids committed:
missing + caused form data to  be wiped to default when saving. Fixed and improved debug level loggig.

flammableliquids committed:
Beginning to improve form field validation presentation to end user

flammableliquids committed:
Custom validator class for webserver IPAddress feild created. Will now check for x.x.x.x and if present it will not stop the user from changing settings.
if the field's data doesn't match x.x.x.x, then pass it to the standard validator

flammableliquids committed:
moved validator object creation into separte function so it's easier for others to read.

flammableliquids committed:
trying to get for validation errors to be more apparent.

flammableliquids committed:
IPAddress_custom inheritence fixed.

validate_umask  and validate_non_manditory_string changed from functions to classes.

Functions for validators don't appear to execute when the validators are created dynamically.

flammableliquids committed:
versioning alignment until github actions works properly

flammableliquids committed:
de-linting

flammableliquids committed:
sonarqube style changes

flammableliquids committed:
matching style and naming conventions based on SonarQube recommendations

flammableliquids committed:
sonar cube recommended moving from a ref to button, so these are style changes for it.

1337-server commented on arm/runui.py:
```diff
@@ -4,7 +4,8 @@
 
 # set the PATH to /arm/arm, so we can handle imports properly
-sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
+# sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
```
Remove the commented out code

flammableliquids committed:
dynamic resize and full width is proably for the best, as it will work with dynamic sizing of page and mobile devices.
Field focus is now highlighted by orange gradient (I like orange)

Attempting to fix version conflict, by incrementing it.

1337-server commented on arm/ui/settings/settings.py:
```diff
@@ -176,27 +188,43 @@ def save_settings():
+            success = True
+        except Exception as e:
+            app.logger.exception(f"Error saving arm.yaml: {e}")
+            flash(f"Error saving arm.yaml: {e}", "error")
```
I thought we moved away from flash() ? And were using toasts/notification system ?

Because the flash() doesn't update till the next page load, so there will be no user feedback until they go to another page. You can see this, by failing to update the settings and then changing the page, a badly formed box will appear. But only after then user has already changed the page.

flammableliquids replied:
I'm not sure if we moved away from it in the past.
I was looking into handling the 'Flash' via ajax.

Your correct throwing the error to the user on form submit is best, waiting for page refresh is ridiculous.
But until I can sort out the ajax end, it's best to leave these in as a sort of 'placeholder', IMHO.

1337-server commented on arm/ui/settings/settings.py:
```diff
@@ -176,27 +188,43 @@ def save_settings():
+        return {'success': success, 'settings': cfg.arm_config, 'form': form_name}
+    else:
+        app.logger.error(f"Error validating form: {form.errors}")
+        flash(f"Error validating form: {form.errors}", "error")
```
See previous, comment.

1337-server commented on arm/ui/settings/templates/settings/ui.html:
```diff
@@ -3,113 +3,14 @@
+                    {% endif %}
+                {% endif %}
+            {% endfor %}
             <button id="uiSubmit" class="btn btn-secondary btn-lg btn-block" form="uiSettings"
```
This no longer submits when the return/enter key is pressed.

Also, the layout is broken and will need the CSS looked at, as all the inputs are not correctly evenly sized

flammableliquids replied:
The functionality to save on enter has not been removed, it's the main method I have been using to test this (CBF scrolling all the way down).

You might find that the arm.yaml your testing against is throwing form based validation errors. 
<img width="1867" height="261" alt="settings_form_error" src="https://github.com/user-attachments/assets/53587cac-6d92-4d88-84a6-f05182ea43f5" />

Those are form validation errors, for the 'template' arm.yaml.

microtechno9000 replied:
Is there an issue with the template or the new parsing of the arm.yaml template/default?

1337-server commented on arm/ui/templates/base.html:
```diff
@@ -10,9 +10,7 @@
-                        <div id="message1" class="alert alert-{{ category if category !=" message
-            " else 'info' }} text-center"
-                             role="alert">
+                        <div id="message1" class="alert alert-{{ category if category != "message" else 'info' }} text-center" role="alert">
```
See comment about flash. These are no longer used as it's better to use the notification system.

flammableliquids replied:
This change was to remove a carriage return that might be the cause of problems when rendering the page.

1337-server reviewed (changes requested):
Still some work to do. I would really advise against using the flash() as this doesn't work well with everything else updating instantly

Return/Enter key should still submit forms.

But the changes to make true/false much simpler is great.

flammableliquids commented:
Based on your testing, and additional re-runs of testing, the validators are flagging problems, which then prevents 'hitting enter' to save. 
I will try and figure out what part broke 'enter to save'

flammableliquids committed:
- commented line removed.
- improved logic for a custom validator.
- all validators removed until client side feedback is available to indicate which field is invalid.
